fairscale/nn/data_parallel/fsdp_optim_utils.py

```diff
@@ -100,7 +117,7 @@ def _unflatten_optim_state(
 
     # non_tensor_state refers to entries in sd[state][param_id] that are not tensors, like "step".
     # we check that these are identical across workers and then take the first
-    non_tensor_state = [_extract_non_tensor_state(combined_state, id) for id in combined_state]
+    non_tensor_state = {id: _extract_non_tensor_state(combined_state, id) for id in combined_state}
```

Comment:
this makes the non_tensor_state a dict with keys since the keys can be non-consecutive, like 0,2. The original code used a list, which keys like 0,1,2 naturally worked.

Reply:
So ids in combined_state can now have None?

Reply:
no, see answers below too. Here is the ID from state_dict is skippy, doesn't have none. The None ID is only used in our own mapping data structure from global to local.